text/0474-event-bridge-scheduler-l2.md

```diff
@@ -0,0 +1,569 @@
+2. **Targets**: A target is an API operation that EventBridge Scheduler calls on your behalf every time your schedule runs. EventBridge Scheduler supports two types of targets: templated targets and universal targets. Templated targets invoke common API operations across a core groups of services. For example, EventBridge Scheduler supports templated targets for invoking AWS Lambda Function or starting execution of Step Function state machine. For API operations that are not supported by templated targets you can use customizeable universal targets. Universal targets support calling more than 6,000 API operations across over 270 AWS services.
+3. **Schedule Group**: A schedule group is an Amazon EventBridge Scheduler resource that you use to organize your schedules. Your AWS account comes with a default scheduler group. A new schedule will always be added to a scheduling group. If you do not provide a scheduling group to add to, it will be added to the default scheduling group. You can create up to 500 schedule groups in your AWS account. Groups can be used to organize the schedules logically, access the schedule metrics and manage permissions at group granularity (see details below). Scheduling groups support tagging: with EventBridge Scheduler, you apply tags to schedule groups, not to individual schedules to organize your resources.
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project. It allows you to define Event Bridge Schedules.
+
+
+## Defining a schedule 
+
+```ts
+const target = new targets.LambdaInvoke(props.func, {
+    input: ScheduleTargetInput.fromObject({
+        "payload": "useful"
+    })
+});
+    
+const schedule = new Schedule(this, 'Schedule', {
+    schedule: ScheduleExpression.rate(Duration.minutes(10)),
+    target,
+    description: 'This is a test schedule that invokes lambda function every 10 minutes.',
+});
+```
+
+### Schedule Expressions
+
+You can choose from three schedule types when configuring your schedule: rate-based, cron-based, and one-time schedules. 
+
+Both rate-based and cron-based schedules are recurring schedules. You configure each recurring schedule type using a schedule expression for the type of schedule you want to configure, and specifying a time zone in which EventBridge Scheduler evaluates the expression with `scheduleTimeZone` parameter.
+
+```ts
+const rateBasedSchedule = new Schedule(this, 'Schedule', {
+    schedule: ScheduleExpression.rate(Duration.minutes(10)),
+    target,
+    description: 'This is a test rate-based schedule',
+});
+
+const cronBasedSchedule = new Schedule(this, 'Schedule', {
+    schedule: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11' }),
+    target,
+    scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,
```

Comment:
How about renaming the prop `scheduleTimezone` to `timezone` to make it short? I feel it is already clear enough that this timezone property is for the schedule.

Reply:
I am concerned that the customer might think that timezone is also applied to other properties. For example, `Schedule` also has properties `startDate` and `endDate`, which should be specified in UTC timezone and do not depend on timezone.

So we have these options: 

1 - Do nothing

Keep interface as-is with a separate `scheduleTimeZone` parameter. It's simple and matches CFN structure https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-scheduler-schedule.html#cfn-scheduler-schedule-scheduleexpressiontimezone

2 - Rename `scheduleTimeZone` to `timeZone` 

Not obvious that the property `timeZone` does not apply to `startDate` and `endDate`

3 -  Add timezone parameter to `ScheduleExpression` methods

Example: 
```ts
const cronBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11', timezone: TimeZone.AMERICA_NEW_YORK}),
    target,
    description: 'This is a test cron-based schedule that will run at 11:00 PM, on day 20 of the month, only in November in New York timezone',
});
```

In implementation we will create a class `ScheduleExpression` that will inherit from [`events.Schedule`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-events.Schedule.html) and will have methods `cron(..)`, `at(..)` and `rate(..)` with timezone as an additional parameter. 

Not a bad option. Maybe `timezone` will not be that discoverable as in other options. 


4 - Create a separate interface  with two parameters `expression` and `timezone`

Example: 

```ts
const cronBasedSchedule = new Schedule(this, 'Schedule', {
    scheduleExpression: {
          expression: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11' }),
          timeZone: TimeZone.AMERICA_NEW_YORK
     },
    target,
    description: 'This is a test cron-based schedule that will run at 11:00 PM, on day 20 of the month, only in November in New York timezone',
});
```

However, I am not sure what name to choose for the new interface as names `Schedule` and `ScheduleExpression` are already taken. Any recommendations? 

---

All in all, I am probably still leaning to the first as it's the simplest option. Curious if you guys also have any opinion on this question @Jacco, @kaizencc  and @comcalvi?

Reply:
That makes sense, thanks! To me, option 3 or 4 look nice and more cohesive, but at the same time option 1 sounds reasonable. I'm fine with whatever decision is made :)

Reply:
I think I like option 4, with `scheduleExpression: { schedule: Schedule, timeZone: TimeZone }`

Reply:
Few things to consider: 

How would the original `events.Schedule` be changed if timezone support is added for `aws-events`. I think it would best added `CronOptions` in CDK since it has no meaning for a `rate` expressions `timeZone` has no meaning. It would be kept separate though because in cloudFormation is would probably not be part of the expression but a separate property like here.

So maybe we should still inherit but add the timeZone parameter from `cron` and `at`.

Unflat interfaces give less good code completion in some languages (like python).

Another thing to consider: what if `aws-events` at some point adds another type of schedule that is unsupported by `aws-scheduler`. We could add an error raising override, but maybe we'd better not inherited?

Maybe implement like this:

```ts
import { Duration, TimeZone } from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';

/**
 * Schedule for EventBridge Schedule
 *
 * Note that rates cannot be defined in fractions of minutes.
 *
 * @see https://docs.aws.amazon.com/scheduler/latest/UserGuide/schedule-types.html
 */
export abstract class ScheduleExpression {
  /**
   * Construct a one-time schedule from a date.
   *
   * @param date The date and time to use. The millisecond part will be ignored.
   */
  public static at(date: Date, timeZone?: TimeZone): ScheduleExpression {
    try {
      const literal = date.toISOString().split('.')[0];
      return new LiteralScheduleExpression(`at(${literal})`, timeZone)
    } catch(e) {
      if (e instanceof RangeError){
        throw new Error('Invalid date');
      }
      throw e;
    }
  }

  public static cron(options: events.CronOptions, timeZone?: TimeZone): ScheduleExpression {
    return new LiteralScheduleExpression(events.Schedule.cron(options).expressionString, timeZone);
  }

  public static rate(duration: Duration): ScheduleExpression {
    return new LiteralScheduleExpression(events.Schedule.rate(duration).expressionString);
  }

  public static expression(expression: string, timeZone?: TimeZone) {
    return new LiteralScheduleExpression(expression, timeZone);
  }

  /**
   * Retrieve the expression for this schedule
   */
   public abstract readonly expressionString: string;

  /**
   * Retrieve the timeZone for this schedule
   */
   public abstract readonly expressionTimeZone?: TimeZone;
}

class LiteralScheduleExpression extends ScheduleExpression {
  constructor(public readonly expressionString: string, public readonly expressionTimeZone?: TimeZone) {
    super();
  }

  public _bind() {}
}
```

Reply:
I like that suggestion @Jacco. It's similar to option 3. I was also thinking about the way to reuse functionality of `events.Schedule` with composition, but couldn't come up with a good enough option like yours. 


I will add it to the RFC. Thanks!

Reply:
Agree on not inheriting for now. The ideal scenario is still to put a generic Schedule into core that can be extended by the 4 resources that have the concept of schedule today. But that's out of scope of this RFC. 

(autoscaling, app-autoscaling, scheduler, events are the 4 i can think of)